### PR TITLE
feat: improve clarinet integrate, clarinet contracts publish and testing harness reliability

### PR DESCRIPTION
**Pull request by lgalabru**
As a short term shortcut, Orchestra is going to be consuming Clarinet as library - I currently think that the correct mid term approach would be to move all the code in charge of `clarinet integrate` in its own crate / lib at some point, which is almost the case with the stacks-devnet-js bindings.
Orchestra is heavily relying on a stable devnet, and so I had to address a bunch of gaps, locally, in order to make progress.
Issues addressed: https://github.com/hirosystems/clarinet/issues/239, https://github.com/hirosystems/clarinet/issues/225, https://github.com/hirosystems/clarinet/issues/38.
With the latest changes, it now takes 30s on my M1 to start a devnet with the counter contract, and 1m40 for a larger protocol that would involve 80-ish contracts (vs 5m and 10m with the current version).
This PR is also improving the contract deployment feature (now under `clarinet contracts publish`), way more robust for deploying large protocols and fixing an issue that was happening with Mainnet deployments.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 name = "clarinet"
 version = "0.27.0"
 authors = ["Ludo Galabru <ludo@hiro.so>"]
-edition = "2018"
+edition = "2021"
```
Comment (obycode) [inline Cargo.toml:5]: +1

**Comment by lgalabru** [comment]
Check https://github.com/hirosystems/clarinet/issues/273

File: src/generate/project.rs [hunk 2]
```diff
@@ -341,9 +341,9 @@ balance = 100_000_000_000_000
 # btc_address: mjSrB3wS4xab3kYqFktwBzfTdPg367ZJ2d
 
 [devnet]
-disable_bitcoin_explorer = true
-# disable_stacks_explorer = true
-# disable_stacks_api = true
+disable_stacks_explorer = true
+disable_stacks_api = true
```
Comment (obycode) [inline src/generate/project.rs:345]: Is this intentional to disable the Stacks API and Explorer by default? I would expect the API is normally used, no?

File: src/indexer/chains/stacks.rs [hunk 3]
```diff
@@ -241,90 +238,136 @@ pub fn get_value_description(raw_value: &str) -> String {
...
+    raw_tx: &str,
+) -> Result<
+    (
+        String,
+        StacksTransaction,
+        StacksTransactionKind,
+        u64,
+        String,
+        Option<String>,
+    ),
```
Comment (obycode) [inline src/indexer/chains/stacks.rs:251]: Might be nice to package this up into a struct, or add comments describing what they are.

**Review by obycode — COMMENTED** [review]
Halfway through! Looks great. Lots of improvements in here! I will continue my review later.

**Review by obycode — CHANGES_REQUESTED** [review]
This is looking great! Just some minor comments/questions.

File: examples/counter/tests/counter_test.ts [hunk 4]
```diff
@@ -53,3 +53,30 @@ Clarinet.test({
         "ST1HTBVD3JG9C05J7HBJTHGR0GGW7KXW28M5JS8QE.plaid-token".expectPrincipal('ST1HTBVD3JG9C05J7HBJTHGR0GGW7KXW28M5JS8QE.plaid-token');
     },
 });
+
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply examples/counter/tests/counter_test.ts:56]: When merging this PR, a new library will be published on denoland.
As soon as we have this library published, we can enable this test, demonstrating how the `beforeContractsDeployment` can be used.
I'll add a todo.

File: src/frontend/cli.rs [hunk 5]
```diff
@@ -79,6 +76,9 @@ enum Contract {
     /// Replicate a third-party contract into this project
     #[clap(name = "fork", bin_name = "fork")]
     ForkContract(ForkContract),
+    /// Publish contracts on chain
+    #[clap(name = "publish", bin_name = "publish")]
+    Publish(Publish),
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/frontend/cli.rs:81]: Good point. I changed this in this PR because I have an incoming collision with `clarinet lambdas publish` (or so) that I'd like to avoid. I'd be ok with keeping `check` and `test` as a root commands for now (would be checking/testing contracts+lambdas), but if you have a strong opinion on this one, I can also move these features as `contracts` subcommands.

File: src/indexer/chains/stacks.rs [hunk 6]
```diff
@@ -241,90 +238,136 @@ pub fn get_value_description(raw_value: &str) -> String {
...
+            if tx.is_mainnet() {
+                sender_conditions.address_mainnet().to_string()
+            } else {
+                sender_conditions.address_testnet().to_string()
+            },
+            Some(if tx.is_mainnet() {
+                sponsor_conditions.address_mainnet().to_string()
+            } else {
+                sponsor_conditions.address_testnet().to_string()
+            }),
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/indexer/chains/stacks.rs:288]: We are not getting the same address twice, we are getting 2 different addresses, the one from the sender, the one from the sponsor, if any.

File: src/indexer/chains/stacks.rs (hunk 3 above)
Comment (lgalabru) [reply src/indexer/chains/stacks.rs:251]: hmmm I'll probably go with comments, as this struct would be dereferenced and destructured a few lines later, for feeding another bigger structure.

File: src/publish/mod.rs [hunk 7]
```diff
@@ -134,70 +138,286 @@ pub fn publish_contract(
...
+
+    let contracts_to_deploy = settings.initial_contracts.len();
+    let node_url = settings.node.clone();
+
+    let per_contract_event_tx_moved = per_contract_event_tx.clone();
+    let deploying_thread_handle = std::thread::spawn(move || {
+        let mut total_contracts_deployed = 0;
+        let stacks_rpc = StacksRpc::new(&node_url);
+
+        while contracts_to_deploy != total_contracts_deployed {
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/publish/mod.rs:232]: Yeah I agree, I'll add some comments, this function grew in complexity as I was testing it and uncovering edge cases.

File: src/types/events.rs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    pub sender: String,
+    pub amount: String,
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+pub struct NFTTransferEventData {
+    #[serde(rename = "asset_identifier")]
+    pub asset_class_identifier: String,
+    #[serde(rename = "value")]
+    pub asset_identifier: String,
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/types/events.rs:32]: The naming that we're using currently (and being corrected / abstracted in this PR) is source of confusion for developers. There was a discussion on the Stacks forum that I'm unable to retrieve.
This new types and schemas, based on Rosetta, are an opportunity to abstract and fix some confusing that we came up with a few months ago.

[the remark this answers is not in the document]
Comment (lgalabru) [reply src/types/events.rs:32]: I remember @friedger being involved in that discussion, maybe he can point us to it.

File: src/publish/mod.rs [hunk 9]
```diff
@@ -134,70 +138,286 @@ pub fn publish_contract(
...
-    }
+        let _ = per_contract_event_tx_moved.send(PublishUpdate::Completed);
+    });
 
-    let mut results = vec![];
+    let results = vec![];
     let mut deployers_nonces = BTreeMap::new();
-    let mut deployers_lookup = BTreeMap::new();
-    let settings = session.settings;
+    let mut deployers_lookup: BTreeMap<String, Account> = BTreeMap::new();
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/publish/mod.rs:323]: Maybe historically, but does not seem to be the case anymore.

File: README.md [hunk 10]
```diff
@@ -136,7 +136,7 @@ phrases and initial balances. Initial balances are in microSTX.
 Clarinet can handle adding a new contract and its configuration to your project with the following command:
 
 ```bash
-$ clarinet contract new bbtc
+$ clarinet contracts new bbtc
```
[the remark this answers is not in the document]
Comment (obycode) [reply README.md:139]: Resolving this because it is discussed further in another commit.

File: src/frontend/cli.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (obycode) [reply src/frontend/cli.rs:81]: Ah, ok. That makes sense. If `check` and `test` will be covering more than just the contracts, then I think keeping them at the top level makes sense.

File: src/indexer/chains/stacks.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (obycode) [reply src/indexer/chains/stacks.rs:288]: Oops 😄

File: src/types/events.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (obycode) [reply src/types/events.rs:32]: Got it. What stops us from changing the field names in these structs to match?

File: examples/counter/tests/counter_test.ts [hunk 11]
```diff
@@ -1,4 +1,4 @@
-import { Clarinet, Tx, Chain, Account, Contract, types } from 'https://deno.land/x/clarinet@v0.13.0/index.ts';
+import { Clarinet, Tx, Chain, Account, Contract, types } from 'https://deno.land/x/clarinet@v0.27.0/index.ts';
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply examples/counter/tests/counter_test.ts:1]: Addressed with latest

File: src/poke/mod.rs [hunk 12]
```diff
@@ -100,16 +100,21 @@ pub fn load_session_settings(
     ];
     settings.initial_deployer = initial_deployer;
     settings.repl_settings = project_config.repl_settings.clone();
+    settings.disk_cache_enabled = true;
+
+    settings.repl_settings.parser_version = settings.repl_settings.parser_version;
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/poke/mod.rs:105]: Addressed with latest

File: node-bindings/lib/types.ts [hunk 13]
```diff
@@ -479,3 +712,97 @@ export enum CoinAction {
...
+export interface StacksContractInterface {
+  /**
+   * List of defined methods
+   * @type {Array<object>}
+   * @memberof ContractInterfaceResponse
+   */
+  functions: Array<object>;
+  /**
+   * List of defined variables
+   * @type {Array<object>}
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/lib/types.ts:730]: Addressed with latest

File: node-bindings/lib/types.ts [hunk 14]
```diff
@@ -124,6 +124,239 @@ export interface StacksTransactionMetadata {
...
+   mutated_contracts_radius: string[];
+  /**
+    *
+    * @type {string[]}
+    * @memberof StacksTransactionReceipt
+    */
+   mutated_assets_radius: string[];
+  /**
+    *
+    * @type {Array<StacksTransactionEvent>}
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/lib/types.ts:186]: Addressed with latest

File: node-bindings/lib/types.ts [hunk 15]
```diff
@@ -479,3 +712,97 @@ export enum CoinAction {
...
+ * @interface DataNftField
+ */
+ export interface DataNftField {
+  /**
+   * Name of non fungible token
+   * @type {string}
+   * @memberof DataNftField
+   */
+  name: string;
+}
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/lib/types.ts:808]: Addressed with latest.
I run Prettier manually from time to time, for now.

File: deno/index.ts [hunk 16]
```diff
@@ -221,19 +226,31 @@ export class Clarinet {
...
+            sessionId: chain.sessionId,
+          }));
+        }
+
         let chain = new Chain(result["session_id"]);
         let accounts: Map<string, Account> = new Map();
         for (let account of result["accounts"]) {
           accounts.set(account.name, account);
-        }
+        }
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply deno/index.ts]: Addressed with latest

File: examples/counter/tests/counter_test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lgalabru) [reply examples/counter/tests/counter_test.ts:56]: Addressed with latest

File: Cargo.toml [hunk 17]
```diff
@@ -67,9 +67,13 @@ tracing-appender = "0.2.0"
 ctrlc = "3.1.9"
 strum = { version = "0.23.0", features = ["derive"] }
 bitcoincore-rpc = "0.14.0"
+bitcoincore-rpc-json = "0.14.0"
 segment = { version = "0.1.2", optional = true }
 mac_address = { version = "1.1.2", optional = true }
 tower-lsp = { version = "0.14.0", optional = true }
+anyhow = "1.0"
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply Cargo.toml]: Addressed with latest

File: src/bin.rs [hunk 18]
```diff
@@ -12,6 +12,9 @@ extern crate rocket;
 #[macro_use]
 extern crate serde_json;
 
+#[macro_use]
+extern crate anyhow;
+
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/bin.rs]: Addressed with latest

File: src/generate/project.rs (hunk 2 above)
Comment (lgalabru) [reply src/generate/project.rs]: Addressed in latest

File: src/indexer/chains/stacks.rs (hunk 3 above)
Comment (lgalabru) [reply src/indexer/chains/stacks.rs]: Addressed with latest

File: src/indexer/mod.rs [hunk 19]
```diff
@@ -108,25 +104,63 @@ impl Indexer {
...
+        let (_, anchored_block) = self.stacks_last_7_blocks.back().unwrap();
+
+        let microblock = chains::standardize_stacks_microblock(
+            &self.config,
+            marshalled_microblock,
+            &anchored_block.block_identifier,
+            &mut self.stacks_context,
+        );
+        self.current_microblock_trail.microblocks.push(microblock);
+        let (_, anchored_block) = self.stacks_last_7_blocks.back().unwrap();
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/indexer/mod.rs]: Addressed with latest

File: src/indexer/chains/stacks.rs [hunk 20]
```diff
@@ -371,21 +414,27 @@ pub fn get_standardized_non_fungible_currency_from_asset_class_id(
 }
 
 pub fn get_standardized_stacks_operations(
-    transaction: &NewTransaction,
+    txid: &str,
+    _transaction: &StacksTransaction,
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/indexer/chains/stacks.rs:418]: Addressed with latest

File: src/integrate/chains_coordinator.rs [hunk 21]
```diff
@@ -501,70 +585,75 @@ pub fn handle_ping() -> Json<JsonValue> {
...
+    use base64::encode;
+    use reqwest::Client;
+
+    // if let Ok(tx_sender) = devnet_events_tx.lock() {
+    //     let _ = tx_sender.send(DevnetEvent::debug(format!(
+    //         "Forwarding request {:?}",
+    //         bitcoin_rpc_call.method
+    //     )));
+    // }
+
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/integrate/chains_coordinator.rs]: Removed in latest

File: src/lib.rs [hunk 22]
```diff
@@ -14,8 +14,13 @@ extern crate serde_json;
 #[macro_use]
 extern crate lazy_static;
 
+#[macro_use]
+extern crate anyhow;
+
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/lib.rs]: Addressed with latest

File: src/publish/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/publish/mod.rs]: Using a hash map in latest

File: src/types/events.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/types/events.rs]: Changing the name of these fields at the stacks-node level would break clients like the stacks-blockchain-api, but also other clients that we don't have control over.

File: src/publish/ui/app.rs [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+    pub fn reset(&mut self) {}
+
+    pub fn display_contract_status_update(&mut self, update: ContractUpdate) {
+        let mut index = 0;
+        for (i, contract) in self.contracts.items.iter().enumerate() {
+            if contract.contract_id == update.contract_id {
+                index = i;
+                break;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/publish/ui/app.rs:45]: Addressed with latest

File: src/integrate/orchestrator.rs [hunk 24]
```diff
@@ -870,31 +800,40 @@ p2p_bind = "0.0.0.0:{}"
...
 retry_count = 255
+include_data_events = true
 events_keys = ["*"]
 "#,
             devnet_config.stacks_node_rpc_port,
             devnet_config.stacks_node_p2p_port,
             devnet_config.miner_secret_key_hex,
             devnet_config.miner_secret_key_hex,
-            format!("bitcoin-node.{}", self.network_name),
+            format!("host.docker.internal"),
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/integrate/orchestrator.rs:833]: Addressed with latest

File: src/lsp/clarity_language_backend.rs [hunk 25]
```diff
@@ -311,63 +311,32 @@ impl ClarityLanguageBackend {
...
+                            .to_str()
+                            .unwrap()
+                            .to_string()
+                    })
+                    .collect::<Vec<_>>();
+                self.client
+                    .show_message(
+                        MessageType::Error,
+                        format!(
+                            "Errors detected in following contracts: {}",
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/lsp/clarity_language_backend.rs]: Addressed with latest, this was actually the result of a bad merge, thank you for catching this.

File: src/publish/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/publish/mod.rs:243]: Addressed with latest

File: examples/counter/tests/counter_test.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (obycode) [reply examples/counter/tests/counter_test.ts:1]: The generator now uses 0.28.0 but this is 0.27.0

File: examples/counter/tests/counter_test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (obycode) [reply examples/counter/tests/counter_test.ts:56]: Ok, makes sense. Thanks

File: src/publish/mod.rs [hunk 26]
```diff
@@ -134,70 +138,297 @@ pub fn publish_contract(
...
-            std::process::exit(1);
+    let (tx, rx) = channel();
+    let (per_contract_event_tx, per_contract_event_rx) = channel();
+
+    let contracts_to_deploy = settings.initial_contracts.len();
+    let node_url = settings.node.clone();
+
+    // Approach's description: after getting all the contracts indexed by the manifest, we build a
+    // a sorted set that we will be splitting in batches of 25 contracts, which is the number of
+    // transactions that you can have for one user at a givent time in a mempool.
```
Comment (obycode) [inline src/publish/mod.rs:231]:
```suggestion
    // transactions that you can have for one user at a given time in a mempool.
```

File: src/indexer/chains/stacks.rs [hunk 27]
```diff
@@ -241,90 +238,136 @@ pub fn get_value_description(raw_value: &str) -> String {
...
-                .call_read_only_fn(
-                    &contract_address,
-                    &contract_name,
-                    "get-decimals",
-                    vec![],
-                    &contract_address,
-                )
-                .expect("Unable to retrieve decimals");
-            let decimals = value.expect_result_ok().expect_u128() as u8;
+            // TODO(lgalabru): re-approach this, with an adequate runtime strategy.
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/indexer/chains/stacks.rs:341]: I'll create an issue for this one, this is still experimental and the solution requires a bit of redesign.

**Review by obycode — APPROVED** [review]
Looks good! Thanks for the fixes.
